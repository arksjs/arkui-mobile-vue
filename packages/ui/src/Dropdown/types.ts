import type { PopupProps, PopupEmits } from '../popup/types'
import type { Selector } from '../helpers'

export interface DropdownProps extends PopupProps {
  selector?: Selector
}

export type DropdownEmits = PopupEmits
