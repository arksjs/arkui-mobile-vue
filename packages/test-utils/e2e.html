<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
