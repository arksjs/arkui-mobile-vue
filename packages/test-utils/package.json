{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
