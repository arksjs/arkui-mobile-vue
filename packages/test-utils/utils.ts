export const timeout = (n: number) => new Promise(r => setTimeout(r, n))
