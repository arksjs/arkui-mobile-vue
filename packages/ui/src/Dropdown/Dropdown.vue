<template>
  <teleport to="body">
    <div
      class="ta-dropdown"
      :class="popupClasses"
      :style="popupStyles"
      v-bind="$attrs"
      ref="popupEl"
    >
      <div class="ta-mask" @click="onMaskClick"></div>
      <div class="ta-dropdown_inner">
        <slot :height="height"></slot>
      </div>
    </div>
  </teleport>
</template>

<script lang="ts">
import {
  defineComponent,
  computed,
  ref,
  nextTick,
  shallowRef,
  type PropType
} from 'vue'
import { usePopup } from '../popup/use-popup'
import { popupEmits, popupProps } from '../popup/props'
import {
  selectorValidator,
  querySelector,
  type PropsToEmits,
  type Selector
} from '../helpers'
import { useException } from '../hooks'
import type { DropdownEmits } from './types'

export default defineComponent({
  name: 'ta-dropdown',
  props: {
    ...popupProps,
    selector: {
      type: [String, HTMLElement] as PropType<Selector>,
      validator: selectorValidator
    }
  },
  emits: { ...popupEmits } as PropsToEmits<DropdownEmits>,
  setup(props, ctx) {
    const { printPropError } = useException()
    const top = ref(-1)
    const height = ref(0)
    const popupEl = shallowRef<HTMLElement | null>(null)

    function updatePos() {
      const $target = querySelector(props.selector)

      if (!$target) {
        printPropError(`Cannot find element through "selector"`)
        return
      }

      const { bottom } = ($target as HTMLElement).getBoundingClientRect()

      top.value = bottom
      nextTick(() => {
        height.value = popupEl.value?.offsetHeight ?? 0
      })
    }

    const popupHook = usePopup(props, ctx, {
      emitCallback(event, res) {
        if (event === 'visibleStateChange') {
          switch (res.state) {
            case 'show':
              updatePos()
              break
            case 'hidden':
              top.value = -1
              break
            default:
              break
          }
        }
      }
    })

    const popupStyles = computed(() => {
      return {
        zIndex: popupHook.zIndex.value,
        top: top.value === -1 ? '100vh' : top.value + 'px'
      }
    })

    return {
      ...popupHook,
      popupStyles,
      top,
      height,
      popupEl
    }
  }
})
</script>
