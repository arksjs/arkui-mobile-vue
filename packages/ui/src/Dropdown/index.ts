import { withInstall } from '../helpers/with-install'
import Dropdown from './Dropdown.vue'

export { Dropdown }
export const install = withInstall(Dropdown)
export default Dropdown
