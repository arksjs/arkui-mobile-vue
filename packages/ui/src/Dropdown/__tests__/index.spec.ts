import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Dropdown } from '@/Dropdown'

describe('Dropdown', () => {
  const selector = 'body'

  test.concurrent('snapshot', () => {
    const wrapper = shallowMount(Dropdown, {
      global: {
        stubs: { teleport: true }
      },
      props: {
        selector
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('slot default', async () => {
    const content = 'content'
    const wrapper = mount(Dropdown, {
      global: {
        stubs: { teleport: true }
      },
      props: {
        selector
      },
      slots: {
        default: content
      }
    })
    expect(wrapper.find('.ak-dropdown_inner').text()).toBe(content)
  })
})
