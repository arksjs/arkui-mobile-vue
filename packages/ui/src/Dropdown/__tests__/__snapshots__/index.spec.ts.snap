// Vitest Snapshot v1

exports[`Dropdown > snapshot 1`] = `
"<teleport-stub>
  <div class=\\"ak-dropdown ak-popup\\" style=\\"z-index: 2000; top: 100vh;\\">
    <div class=\\"ak-mask\\"></div>
    <div class=\\"ak-dropdown_inner\\"></div>
  </div>
</teleport-stub>"
`;
